### DISCUSSION TITLE
Upgrades issue

### DISCUSSION THREAD

**Discussion by saikek** [arendst/Tasmota#12286] [Q&A] [answered]
I have almost 15 Tasmota devices on my network.

I've been doing rolling update from ~6.2.0 to 9.4.0.4 and for some of them was getting Error 29 - if my memory serves me correctly. (something loop-related)

I've tried both OTA (using local node server), web upload, upload through MQTT commands (most stable).

But 2 devices refuses to update at all, stuck with minimal configs.

I found out that reason was Linux... Just by doing web upload using `Edge & Windows` went without any issues, from first attempt. Firewall was off on Linux, so I don't know what was the issue.

So just a solution for anyone using Linux (Kubuntu 20.4 in my case) and would have same issue.

**Comment by sfromis** [comment, marked as answer]
Going directly from 6.2 to 9.4 (not sure what you mean by "rolling") is a good recipe for getting bad errors, including bricked devices, unless you follow the long stepwise procedure on [Upgrading](https://tasmota.github.io/docs/Upgrading/). If (or rather when) a device does not work properly after that, you could see if you're able to do a `Reset 1` or [Recovery](https://tasmota.github.io/docs/Device-Recovery/), but that may not be feasible. Issues here is more likely than Kubuntu being the culprit. If you have any chance of wired serial flashing, I'd suggest that as quicker than trying all sorts of things to make it work.

I'm running Ubuntu Mate, no problems at all, as long as I install the right software, and grant myself access to serial (if using that).

**Reply by saikek** [reply]
Yes, I went exactly through that process, 
6.2 -> 6.2 minimal -> 6.7.1 minimal -> 7.20 minimal -> 8.5.1 minimal -> 9.1 minimal.gz -> 9.1

I've done `reset 6` before flashing.

It was just frustrating, that some devices were stuck on same version, without ability to change anything.
Serial would not work, since they were wall light switches, which would take too much effort to flash.

**Reply by sfromis** [reply]
Does that mean that you got stuck upgrading from 9.1 tasmota-minimal to full? Did you use the .gz binary also for full, tasmota-bin.gz?

**Reply by saikek** [reply]
I've tried all combinations - .gziped and not, different types of firmware. That was my point - not Tasmota fault. 
Also I tried Firefox & Chrome browser on Linux (so different engines).

Not sure how to pinpoint that issue more precisely. Is there some kind of stacktrace that can be printed ?

**Reply by barbudor** [reply]
> 6.2 -> 6.2 minimal -> 6.7.1 minimal -> 7.20 minimal -> 8.5.1 minimal -> 9.1 minimal.gz -> 9.1

This is BAD
Any minimal MUST be followed by a full tasmota or tasmota-lite of the SAME version before going to the next version.
The "Upgrading" doc say that multiple time and in different ways to be sure everyone gets it.
Recommendation is to use tamota-lite, not tasmota-minimal.

**Comment by saikek** [comment]
I was not able to find similar issues either in tickets or on google. 

It's just an idea to try for someone who would stumble upon the same issue for search to index.

**Reply by corbolais** [reply]
While the devices at hand did not loop as you suggested yours are, it is equally impossible to upgrade.
I flashed 9.1 minimal on 8.5.1 and then 9.5 minimal on top of that (8.5 -> 9.1 minimal -> 9.5 minimal). Starting with 9.5 I cannot even reflash 9.5 minimal, be it compressed or not, self-compiled or stock.

Issues: https://github.com/arendst/Tasmota/issues/12568, https://github.com/arendst/Tasmota/issues/12575

Discussion: https://github.com/arendst/Tasmota/discussions/12579

**Reply by ascillato2** [reply]
Please, check the upgrading summary:

https://github.com/arendst/Tasmota/discussions/12579#discussioncomment-970413]

Thanks.

**Reply by timaseth** [reply]
To add to that: In understanding the reason of using `minimal` lies the key to avoiding any minimal-related accidents. Minimal/Lite may confuse some and there is merit in reading through the very well written docs site (which is always open to edits and improvements). If found wanting, a PR is encouraged to improve the docs thereby helping the community.

**Comment by ascillato2** [comment]
## Summary:

For upgrading, please, do as following:

### Please, Don't flash minimal as the first firmware.

### Please, don't update from an old version using minimal to minimal.

### Please, follow the upgrade steps https://tasmota.github.io/docs/Upgrading/

### If you see the message **_"upload buffer miscompare"_** means not enough space. Please, use minimal or lite and then use the full version you want.

**_(Not following these steps can end up with a non-working device and you will need to erase all flash and flash by serial)_**

### UPGRADE OTAurl WAY:

**Please, don't use minimal in OTAurl**. Tasmota will take care of downloading and upgrading the minimal version automatically in-between updates. Just point to the Full version you want.

+ Upgrade from OLD Full Version -> New Full Version
   + Example: From Full 8.5.1 -> Full 9.1 -> 9.5 FULL
          + Set the following OTAurl in Tasmota UI and click on OTAUrl Upgrade
                http://ota.tasmota.com/tasmota/release-9.1.0/tasmota.bin.gz
          + (WAIT 2 to 5 minutes and don't reset the device)
          + Set the following OTAurl in Tasmota UI and click on OTAUrl Upgrade
                http://ota.tasmota.com/tasmota/tasmota.bin.gz
          + (WAIT 2 to 5 minutes and don't reset the device)

### UPGRADE MANUAL WAY:

**Please, don't update from an old version using minimal to minimal.**
+ Upgrade from OLD Full Version -> New minimal version -> New Full Version
   + Example: Full 8.5.1 -> minimal 9.1 -> Full 9.1 -> 9.5 minimal -> 9.5 FULL
+ Alternatively: Full 8.51 -> minimal 9.1 -> Lite 9.1 -> Lite 9.5

_(For more information on this topic, please refer to the Tasmota Docs at https://tasmota.github.io/docs/Upgrading/ and older issues. Thanks.)_
